packages/navigator/src/browser/navigator-editor-sync.ts

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable, inject, postConstruct } from "inversify";
+import { Disposable } from '@theia/core/lib/common';
+import { EditorManager, EditorWidget } from "@theia/editor/lib/browser";
+import { ITreeNode, ISelectableTreeNode, IExpandableTreeNode, PreferenceChangeEvent } from "@theia/core/lib/browser";
+import { FileNavigatorModel } from "./navigator-model";
+import { NavigatorConfiguration, NavigatorPreferences } from "./navigator-preferences";
+
+@injectable()
+export class NavigatorEditorSynchronizer {
+
+    @inject(FileNavigatorModel)
+    private readonly fileNavigatorModel: FileNavigatorModel;
+
+    @inject(EditorManager)
+    private readonly editorManager: EditorManager;
+
+    @inject(NavigatorPreferences)
+    private readonly navigatorPreferences: NavigatorPreferences;
+
+    private activeEditorChangedSubscription: Disposable | undefined;
+
+    @postConstruct()
+    protected async init() {
+        await this.navigatorPreferences.ready;
+        this.linkWithEditorPreferenceChanged(this.navigatorPreferences['navigator.linkWithEditor']);
+        this.navigatorPreferences.onPreferenceChanged(preference => this.onPreferenceChangedHandler(preference));
+    }
+
+    private onPreferenceChangedHandler(preference: PreferenceChangeEvent<NavigatorConfiguration>) {
+        if (preference.preferenceName === 'navigator.linkWithEditor') {
+            this.linkWithEditorPreferenceChanged(preference.newValue);
+        }
+    }
+
+    private linkWithEditorPreferenceChanged(linkWithEditorNewValue: boolean | undefined) {
+        if (linkWithEditorNewValue) {
+            if (this.activeEditorChangedSubscription === undefined) {
+                this.activeEditorChangedSubscription = this.editorManager.onActiveEditorChanged(
+                    (editor: EditorWidget | undefined) => this.selectNodeByEditor(editor));
+            }
+        } else {
+            if (this.activeEditorChangedSubscription !== undefined) {
+                this.activeEditorChangedSubscription.dispose();
+                this.activeEditorChangedSubscription = undefined;
+            }
+        }
+    }
+
+    /**
+     * Reveals and selects corresponding to given editor node in the navigator.
+     * If editor is undefined nothing happens.
+     *
+     * @param editor editor widget object to reveal its node in the navigator
+     */
+    selectNodeByEditor(editor: EditorWidget | undefined) {
+        if (editor) {
+            this.selectNodeById(this.editorIdToNavigatorNodeId(editor.id));
```

Comment:
You should access an uri of the underlying model and use it as an id

Reply:
@akosyakov @svenefftinge @kittaakos just to discuss: we might have different types of editors with different underlying models, each of them has its own structure. Seems there is no unique way to get the uri from underlying model (which should be used as an id). So, maybe we should add explicit contract (an interface) which shows that a widget is related to a _physical file_ somehow (we can use uri in the contract as well). The difference between just an uri field and the contract is that _just an uri_ could point to any type of resource, for example a web page. That's could be a problem. At least cause running some logic without result (error could be suppressed, but still). And this explicit contract might be (and will be if so) used in navigator without hard dependency on any editor/viewer.

Reply:
> maybe we should add explicit contract (an interface)

If you would like to go with nominal typings, you would have to depend on the `navigator` extension from the `editor` and the other contributing extensions.

> The difference between just an uri field and the contract is that just an uri could point to any type of resource, for example a web page.

Exactly, I have a counterexample, you have a diff editor which does not necessarily have a `file` schema, but you could have a `uri` property getter where you return with the `file` URI; hence you can reveal it in the navigator.

> And this explicit contract might be (and will be if so) used in navigator without hard dependency on any editor/viewer.

An explicit contract can be documentation that says if you would like to contribute to this feature, have a public `uri` with type `URI`.

Long story short, I like the duck typing approach, but I am OK with anything else unless we do not introduce any dependencies between extensions.

Reply:
- We should not use `uri` property name. `uri` should point to a real URI, not a URI computed based on a real. As @kittaakos mentioned an example is the diff editor widget:
  - a real is a diff URI
  - a target URI is computed based on it
- I am not sure that it is always possible to compute the target URI. It should be optional.
- I am in favor of the type safety, that's the point why we use ts.

Let's introduce `Navigatable` interface in the core:
```ts
export interface Navigatable {
    readonly targetUri: URI | undefined;
}
export namespace Navigatable {
    export function is(arg: Object | undefined): arg is Navigatable {
        return !!arg && 'targetUri' in arg;
    }
}
```

- The editor widget needs to take care of a case when an underlying editor URI is a diff URI.
- The navigator should use it.

Reply:
> I am in favor of the type safety, that's the point why we use TypeScript.

Duck typing does not mean it is not typesafe.

Reply:
ok, I meant it should be possible to find all implementors, rename properties knowing that you don't break any contracts and so on, tooling should be aware of it and help.